File: Client/src/lib/api/services/authService.ts
import apiClient from "../apiClient";
import { type User } from "@/store/useAuthStore";
import { v7 as uuidv7 } from "uuid";

export interface ApiResponse<T> {
   message: string;
   errorCode?: string;
   data: T;
   requestId?: string;
}

export interface LoginRequest {
   email: string;
   password: string;
   rememberMe?: boolean;
}

export interface RegisterRequest {
   email: string;
   password: string;
   fullName: string;
   rememberMe?: boolean;
}

let refreshPromise: Promise<User> | null = null;

export const authService = {
   /**
    * Authenticates a user with email and password.
    */
   login: async (payload: LoginRequest): Promise<User> => {
      const response = (await apiClient.post("/auth/login", payload, {
         withCredentials: true,
         // @ts-expect-error - Custom axios config flag for interceptor
         _skipToast: true,
      })) as unknown as ApiResponse<User>;
      return response.data;
   },

   /**
    * Authenticates a user with 2FA OTP code.
    */
   loginWith2FA: async (payload: {
      email: string;
      password: string;
      code: string;
      rememberMe?: boolean;
   }): Promise<User> => {
      const response = (await apiClient.post("/auth/2fa/login", payload, {
         withCredentials: true,
         // @ts-expect-error - Custom axios config flag for interceptor
         _skipToast: true,
      })) as unknown as ApiResponse<User>;
      return response.data;
   },

   /**
    * Registers a new user.
    */
   register: async (payload: RegisterRequest): Promise<ApiResponse<User>> => {
      const idempotencyKey = uuidv7();
      const response = (await apiClient.post("/auth/register", payload, {
         withCredentials: true,
         headers: {
            "X-Idempotency-Key": idempotencyKey,
         },
         // @ts-expect-error - Custom axios config flag for interceptor
         _skipToast: true,
      })) as unknown as ApiResponse<User>;
      return response;
   },

   /**
    * Verifies user email with a token.
    */
   verifyEmail: async (token: string) => {
      return apiClient.get<ApiResponse<void>>(`/auth/verify?token=${token}`);
   },

   /**
    * Resends the verification email to the logged-in user.
    */
   resendVerification: async () => {
      return apiClient.post<ApiResponse<void>>("/auth/resend-verification");
   },

   /**
    * Refreshes the current session and returns latest user data.
    * Uses a promise cache to prevent parallel refresh requests.
    */
   refresh: async (): Promise<User> => {
      if (refreshPromise) return refreshPromise;

      refreshPromise = (async () => {
         try {
            const response = (await apiClient.post(
               "/auth/refresh",
               {},
               {
                  withCredentials: true,
                  // @ts-expect-error - Custom axios config flag for interceptor
                  _skipToast: true,
                  _skipRedirect: true,
               }
            )) as unknown as ApiResponse<User>;
            return response.data;
         } finally {
            refreshPromise = null;
         }
      })();

      return refreshPromise;
   },

   /**
    * Logs out the current user.
    */
   logout: async () => {
      return apiClient.post<ApiResponse<void>>("/auth/logout", {}, { withCredentials: true });
   },

   /**
    * Requests a password reset email.
    */
   forgotPassword: async (email: string) => {
      return apiClient.post<ApiResponse<void>>("/auth/forgot-password", { email });
   },

   /**
    * Resets the password using a token.
    */
   resetPassword: async (payload: { token: string; newPassword: string; logoutAll?: boolean }) => {
      return apiClient.post<ApiResponse<void>>("/auth/reset-password", payload);
   },

   /**
    * Gets the OAuth2 Auth URL for a specific provider.
    */
   getOAuth2Url: async (provider: string): Promise<ApiResponse<string>> => {
      const response = (await apiClient.get(`/auth/${provider}/url`, {
         withCredentials: true,
      })) as unknown as ApiResponse<string>;
      return response;
   },

   /**
    * Finalizes registration session by setting HttpOnly cookies.
    */
   finalizeRegistration: async (accessToken: string): Promise<User> => {
      const response = (await apiClient.post(
         "/auth/finalize-registration",
         { accessToken },
         {
            withCredentials: true,
            // @ts-expect-error - Custom axios config flag for interceptor
            _skipToast: true,
         }
      )) as unknown as ApiResponse<User>;
      return response.data;
   },
};

File: Server/src/main/java/com/omnibooking/services/auth/AuthService.java
package com.omnibooking.services.auth;

import com.omnibooking.dto.AuthResponse;
import com.omnibooking.dto.LoginRequest;
import com.omnibooking.dto.RegisterRequest;
import com.omnibooking.dto.TwoFactorLoginRequest;
import com.omnibooking.dto.oauth.OAuth2UserInfo;
import com.omnibooking.security.RedisSessionInfo;

import java.util.UUID;
import jakarta.servlet.http.HttpServletResponse;

public interface AuthService {

   AuthResponse register(RegisterRequest request, String ip, String userAgent, HttpServletResponse response,
         boolean rememberMe);

   AuthResponse login(LoginRequest request, String ip, String userAgent, HttpServletResponse response, String oldSessionId);

   AuthResponse loginWith2FA(TwoFactorLoginRequest request, String ip, String userAgent, HttpServletResponse response, String oldSessionId);

   AuthResponse loginWithOAuth2(String provider, OAuth2UserInfo userInfo, String ip,
         String userAgent, HttpServletResponse response, boolean rememberMe, String oldSessionId);

   AuthResponse refresh(String sessionId, String refreshToken, String ip, String userAgent,
         HttpServletResponse response);

   void logout(UUID sessionId, UUID userId, HttpServletResponse response);

   void verifyEmail(String token);

   void resendVerification(UUID userId);

   void clearAllCookies(HttpServletResponse response);

   AuthResponse upgradeToPartner(UUID userId, String ip, String userAgent, HttpServletResponse response,
         boolean rememberMe, String oldSessionId);

   void forgotPassword(String email);

   void resetPassword(String token, String newPassword, boolean logoutAll);

   RedisSessionInfo getSessionInfo(String sessionId);

   AuthResponse finalizeRegistration(String accessToken, String ip, String userAgent, HttpServletResponse response, String oldSessionId);

   boolean checkEmail(String email);

   AuthResponse activateGuest(String token, String password, String ip, String userAgent, HttpServletResponse response, String oldSessionId);

   com.omnibooking.dto.RegistrationStatusResponse getRegistrationStatus(String requestId);

}
